File: py/search/array/InterpolationSearch.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class InterpolationSearch:
    def search(self, nums: List[int], target: int) -> int:
        start, end = 0, len(nums) - 1
        while start < end and nums[start] != nums[end]:
            middle = start + (target - nums[start]) * (end - start) // (nums[end] - nums[start])
            if middle < start or middle > end:
                return -1

            if target == nums[middle]:
                return middle

            if target < nums[middle]:
                end = middle - 1
            else:
                start = middle + 1
        return start if start < len(nums) and target == nums[start] else -1

File: java/search/array/InterpolationSearch.java
package search.array;

public class InterpolationSearch {
    public int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;
        while (start < end && nums[start] != nums[end]) {
            int middle = start + (target - nums[start]) * (end - start) / (nums[end] - nums[start]);
            if (middle < start || middle > end) {
                return -1;
            }

            if (target == nums[middle]) {
                return middle;
            }

            if (target < nums[middle]) {
                end = middle - 1;
            } else {
                start = middle + 1;
            }
        }
        return (start < nums.length && target == nums[start]) ? start : -1;
    }
}
